src/test/java/org/opensearch/security/dlic/rest/api/ConfigUpgradeApiActionUnitTest.java

```diff
@@ -0,0 +1,291 @@
+            source.set("c", objectMapper.createObjectNode());
+        }, updated -> {
+            updated.set("a", objectMapper.createObjectNode());
+            updated.set("c", objectMapper.createObjectNode());
+            updated.set("b", objectMapper.createObjectNode());
+        }, List.of()));
+
+        testCases.add(new OperationTestCase("Moved and different object", source -> {
+            source.set("a", objectMapper.createObjectNode());
+            source.set("b", objectMapper.createObjectNode());
+            source.set("c", objectMapper.createObjectNode());
+        }, updated -> {
+            updated.set("a", objectMapper.createObjectNode());
+            updated.set("c", objectMapper.createObjectNode().put("d", "1"));
+            updated.set("b", objectMapper.createObjectNode());
+        }, List.of(List.of("add", "/c/d", "1"))));
+
+        testCases.add(new OperationTestCase("Removed field object", source -> {
+            source.set("a", objectMapper.createObjectNode().put("hidden", true));
+            source.set("b", objectMapper.createObjectNode());
+        }, updated -> {
+            updated.set("a", objectMapper.createObjectNode());
+            updated.set("b", objectMapper.createObjectNode());
+        }, List.of(List.of("remove", "/a/hidden", "<NULL>"))));
+
+        testCases.add(new OperationTestCase("Removed field object", source -> {
+            final var roleA = objectMapper.createObjectNode();
+            roleA.putArray("cluster_permissions").add("1").add("2").add("3");
+            source.set("a", roleA);
+        }, updated -> {
+            final var roleA = objectMapper.createObjectNode();
+            roleA.putArray("cluster_permissions").add("2").add("11").add("3").add("44");
+            updated.set("a", roleA);
+        },
+            List.of(
+                List.of("remove", "/a/cluster_permissions/0", "<NULL>"),
+                List.of("add", "/a/cluster_permissions/1", "11"),
+                List.of("add", "/a/cluster_permissions/3", "44")
+            )
+        ));
+
+        for (final var tc : testCases.build()) {
+            // Setup
+            final var source = objectMapper.createObjectNode();
+            source.set("_meta", objectMapper.createObjectNode().put("type", CType.ROLES.toLCString()).put("config_version", 2));
+            tc.sourceChanges.accept(source);
+            final var updated = objectMapper.createObjectNode();
+            tc.updates.accept(updated);
+
+            var sourceAsConfig = SecurityDynamicConfiguration.fromJson(objectMapper.writeValueAsString(source), CType.ROLES, 2, 1, 1);
+
+            doReturn(ValidationResult.success(sourceAsConfig)).when(configUpgradeApiAction)
+                .loadConfiguration(any(), anyBoolean(), anyBoolean());
+            doReturn(updated).when(configUpgradeApiAction).loadConfigFileAsJson(any());
+
+            // Execute
+            var result = configUpgradeApiAction.computeDifferenceToUpdate(CType.ACTIONGROUPS);
```

Comment:
should this be CType.ROLES?

Reply:
In this unit test any CType is viable - in the future we could support upgrading of any config types in this way.